### PR TITLE
[CMake] Refactor gqcp_version to a static library instead of header-only.

### PR DESCRIPTION
**Pull request by dariatols**
**Short description**

Add gqcp_version with compiled source file instead of header file. This way, `gqcp` (and `gqcpy`) does not need to be recompiled if the version should be modified.

**To do**

- [x] (if applicable, when creating new source files): don't forget to update the Doxygen file, the collective `gqcp.hpp` include header and the CMake files
- [x] Add other smaller task to be completed as a Markdown task list

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -10,18 +10,21 @@ cmake_minimum_required(VERSION 3.13 FATAL_ERROR)
 project(GQCP VERSION 0.0.1 LANGUAGES CXX)
 list(APPEND CMAKE_MODULE_PATH ${CMAKE_SOURCE_DIR}/cmake)  # make CMake look into the ./cmake/ folder for configuration files
 
+# Add the versioning as a separate library. Whenever the version changes, only the version.cpp source file will be recompiled 
+# and only gqcp needs to be relinked. This way, gqcp does not need to be recompiled if the version should be modified.
 # Add configuration file containing version of current library.
 configure_file(${CMAKE_SOURCE_DIR}/cmake/version.hpp.in version/version.hpp)
+configure_file(${CMAKE_SOURCE_DIR}/cmake/version.cpp.in version/version.cpp)
 # Compile the version as its own library where, if needed, other targets can be linked to.
 # Common practice: this to avoid unneccessary recompiling whenever version details change.
```
Comment (lelemmen) [inline CMakeLists.txt:19]: This documentation seems superfluous, given the documentation that you added a couple of lines above it.

File: cmake/version.hpp.in [hunk 2]
```diff
@@ -27,16 +27,50 @@ namespace ${PROJECT_NAME} {
...
+    /**
+     *  @return The full GQCP version.
+     */
+    static std::string full_version();
 
-    static constexpr auto git_SHA1 = "${GIT_SHA1}";
+    /**
+     *  @return The current GQCP commit SHA1.
+     */
+    static std::string git_SHA1();
```
Comment (lelemmen) [inline cmake/version.hpp.in:73]: The code style in this example isn't the same as the rest of the GQCP library:
- We don't start our private variables with an underscore;
- Methods are always in camelCase.

File: gqcpy/version.cpp [hunk 3]
```diff
@@ -21,28 +21,40 @@
...
+
+        .def(
+            "patch",
+            &Version::patch_version,
+            "The GQCP patch version.")
+
+        .def(
+            "full",
+            &Version::full_version,
+            "The full GQCP version.")
```
Comment (lelemmen) [inline gqcpy/version.cpp:52]: Why do you opt for a disconnect between names in the C++ implementation and the Python bindings?

File: cmake/version.hpp.in (hunk 2 above)
Comment (dariatols) [reply cmake/version.hpp.in:73]: @lelemmen I have added the underscores since the variables `major`, `minor` and `patch` are already predefined by GNU.

Comment (dariatols) [reply cmake/version.hpp.in]: The method names have been adjusted according to GQCP style.

File: CMakeLists.txt [hunk 4]
```diff
@@ -25,7 +24,7 @@ target_include_directories(gqcp_version
         $<BUILD_INTERFACE:${CMAKE_BINARY_DIR}/version>
 )
 
-# List the supported options
+# List the supported optionsc
```
Comment (lelemmen) [inline CMakeLists.txt:27]:
```suggestion
# List the supported options.
```

File: gqcpy/__init__.py.in [hunk 5]
```diff
@@ -1,5 +1,4 @@
-from .gqcpy import *
+from.gqcpy import*
```
Comment (lelemmen) [inline gqcpy/__init__.py.in:1]:
```suggestion
from .gqcpy import *
```

File: gqcpy/__init__.py.in [hunk 6]
```diff
@@ -1,5 +1,4 @@
...
 
-
-# Print version-related information upon import
-print("gqcpy version {} @ {}".format(gqcpy.Version.full, gqcpy.Version.git_sha1))
+#Print version - related information upon import
+print("gqcpy version {} @ {}".format(gqcpy.Version.fullVersion(), gqcpy.Version.gitSHA1()))
```
Comment (lelemmen) [inline gqcpy/__init__.py.in:4]:
```suggestion
# Print version-related information upon import.
print("gqcpy version {} @ {}".format(gqcpy.Version.fullVersion(), gqcpy.Version.gitSHA1()))
```

**Review by lelemmen — APPROVED** [review]
Good to go after small adjustments.

**Pull request merged**